File: 51-n-queens/n-queens.py
class Solution:
    def solveNQueens(self, n: int) -> List[List[str]]:
        def convert(board):
            sol=[]
            for col in board:  #[1,3,0,2]
                row=['.']*n
                row[col]='Q' # ['.','Q','.','.']
                sol.append(''.join(row))
            return sol


        def is_valid(x2,y2,board):
            for x1 in range(x2):
                if y2==board[x1] or abs(x1-x2)==abs(board[x1]-y2):
                    return False # not a valid placement
            return True 
        #------------------------------------------
        def backtracking(board,row):
            if row==n:
                res.append(convert(board[:])) # adds a copy of board
                return
            
            for col in range(0,n): # col => 0 ,1,.... n-1
                if is_valid(row,col,board):
                    board[row]=col # make the move
                    backtracking(board,row+1) # explore the path
                    board[row]=-1 # undo the move

        #---------------------------------------
        res=[]                     #.row    0. 1. 2. 3.  ==> board[row]=> col
        board=[-1]*n # n=4.   board=[-1,-1,-1,-1] # col locations of Queens
        backtracking(board,0) # row=0. board=[-1,-1,-1,-1]
        return res

File: 51-n-queens/n-queens.java
class Solution {
    public boolean is_valid(int[] board,int x2,int y2){
        for(int x1=0;x1<x2;x1++){
            int y1=board[x1];
            if(y1==y2 || Math.abs(x1-x2)==Math.abs(y1-y2)){
                return false;
            }
        }
        return true;
    }
    public void backtracking(int n, int[] board,List<List<String>> res,int row){
        if(row==n){ //board[1,3,0,2];
            ArrayList<String> copy=new ArrayList<>();
            for(int i=0;i<n;i++){
                String row_list="";
                for(int j=0;j<n;j++){
                    if(board[i]==j){
                       row_list+="Q";
                    }
                    else{
                       row_list+=".";
                    }
                }
                copy.add(row_list);
            }
            res.add(copy);
            return;
        }

        for(int col=0;col<n;col++){
            if(is_valid(board,row,col)==true){
                board[row]=col; //make the move
                backtracking(n,board,res,row+1);// may lead to a solution
                board[row]=-1;  //undo the move
            }
        }
    }

    public List<List<String>> solveNQueens(int n) {
        List<List<String>> res= new ArrayList<>();
        int[] board=new int[n];
        Arrays.fill(board,-1);

        backtracking(n,board,res,0);

        return res;

    }
}